(new core) Record-valued ValueType, with canonical-bytes enforcement (#1251)
garden-co/jazz: homepage/homepage/app/(docs)/docs/[framework]/[...slug]/authentication/overview.mdx

export const metadata = { title: "Authentication methods" };

import { CodeGroup, ContentByFramework } from "@/components/forMdx";

# Authentication in Jazz

> Jazz authentication is based on cryptographic keys that are stored locally on the user device.

Author: ```suggestion
Jazz authentication is based on cryptographic keys ("Account keys"). Their public part represents a user's identity, their secret part lets you act as that user.
```

> When a user loads a Jazz application for the first time, we generate those keys and store them locally.

Author: ```suggestion
When a user loads a Jazz application for the first time, we create a new Account by generating keys and storing them locally.
```

Without any additional steps the user can use Jazz normally, but they would be limited to use on only one device.

> For an account to use multiple devices, and back up their credentials, you can set up an authentication method using one of the Jazz providers.

Author: To make Accounts work across devices, you can store/retrieve the account keys from an *authentication method* by using the corresponding hooks and providers.

> For an account to use multiple devices, and back up their credentials, you can set up an authentication method using one of the Jazz providers.

Reviewer: ```suggestion
To make Accounts work across devices, you can store/retrieve the account keys from an authentication method by using the corresponding hooks and providers.
```

<ContentByFramework framework={["react", "vue", "svelte"]}>
## Authentication with passkeys

> Passkey authentication is fully local-first and the most secure of the auth methods that Jazz provides.

Author: ```suggestion
Passkey authentication is fully local-first and the most secure of the auth methods that Jazz provides (because keys are manageed by the device/operating system itself).
```

> It is based on the [Web Authentication API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API) and is both very easy to use and widely supported.

Author: ```suggestion
It is based on the [Web Authentication API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API) and is both very easy to use (using familiar FaceID/TouchID flows) and widely supported.
```

<ContentByFramework framework="react">
Using passkeys in Jazz is as easy as this:
<CodeGroup>
{/* prettier-ignore */}
```tsx
export function AuthModal({ open, onOpenChange }: AuthModalProps) {
  const [username, setUsername] = useState("");

  const auth = usePasskeyAuth({ // Must be inside the JazzProvider!
    appName: "My super-cool web app",
  });

  if (auth.state === "signedIn") { // You can also use `useIsAuthenticated()`
    return <div>You are already signed in</div>;
  }

  const handleSignUp = async () => {
    await auth.signUp(username);
    onOpenChange(false);
  };

  const handleLogIn = async () => {
    await auth.logIn();
    onOpenChange(false);
  };

  return (
    <div>
      <button onClick={handleLogIn}>Log in</button>
      <input type="text" value={username} onChange={(e) => setUsername(e.target.value)} />
      <button onClick={handleSignUp}>Sign up</button>
    </div>
  );
}
```
</CodeGroup>
</ContentByFramework>

You can try our passkey authentication using our [passkey example](https://passkey-demo.jazz.tools/) or the [music player demo](https://music-demo.jazz.tools/).

</ContentByFramework>

## Passphrase-based authentication

> Passphrase authentication gives the user a passphrase to log into another device.

Author: ```suggestion
Passphrase authentication lets users log into any device using a Bitcoin-style passphrase. This means users are themselves responsible for storing the passphrase safely.
```

  const handleLogIn = async () => {
    await auth.logIn(loginPassphrase);
    onOpenChange(false);
  };

  return (
    <View>
      <View>
        <Text>Your current passphrase</Text>
        <TextInput
          editable={false}
          value={auth.passphrase}
          multiline
          numberOfLines={5}
        />
      </View>

      <Button
        title="I have stored my passphrase"
        onPress={handleSignUp}
      />

      <View>
        <Text>Log in with your passphrase</Text>
        <TextInput
          value={loginPassphrase}
          onChangeText={setLoginPassphrase}
          placeholder="Enter your passphrase"
          multiline
          numberOfLines={5}
          required
        />
      </View>

      <Button
        title="Log in"
        onPress={handleLogIn}
      />
    </View>
  );
}
```
</CodeGroup>
</ContentByFramework>

You can try our passphrase authentication using our [passphrase example](https://passphrase-demo.jazz.tools/) or the [todo list demo](https://todo-demo.jazz.tools/).

<ContentByFramework framework={["react", "react-native"]}>
## Integration with Clerk

Jazz can be used with [Clerk](https://clerk.com/) to authenticate users.

> This authentication method is not fully local-first, because the login and signup need to be done while online.

Author: ```suggestion
This authentication method is not fully local-first, because the login and signup need to be done while online. Clerk and anyone who is an admin in the app's Clerk org are trusted with the user's key secret and could impersonate them.
```

<CodeGroup>
{/* prettier-ignore */}
```tsx
import { SignInButton } from "@clerk/clerk-react";
import { useAccount, useIsAuthenticated } from "jazz-react";

export function AuthButton() {
  const { logOut } = useAccount();

  const isAuthenticated = useIsAuthenticated();

  if (isAuthenticated) {
    return <button onClick={() => logOut()}>Logout</button>;
  }

  return <SignInButton />;
}
```
</CodeGroup>
</ContentByFramework>

<ContentByFramework framework="react-native">
Then you can use the [Clerk auth methods](https://clerk.com/docs/references/expo/overview) to log in and sign up:
<CodeGroup>
{/* prettier-ignore */}
```tsx
import { SignInButton } from "@clerk/clerk-expo";
import { useAccount, useIsAuthenticated } from "jazz-react-native";

export function AuthButton() {
  const { logOut } = useAccount();
  const { signIn, setActive, isLoaded } = useSignIn();

  const isAuthenticated = useIsAuthenticated();

  if (isAuthenticated) {
    return <button onClick={() => logOut()}>Logout</button>;
  }

  // Login code with Clerk Expo
}
```
</CodeGroup>
</ContentByFramework>
</ContentByFramework>

## Migrating data from anonymous to authenticated account

> You may want allow your users to use your app without authenticating (a poll response for example). But they might make changes that they'd like to keep when they do authenticate.

Author: ```suggestion
You may want allow your users to use your app without authenticating (a poll response for example). When *signing up* their anonymous account is transparently upgraded using the provided auth method, keeping the data stored in the account intact.

However, a user may realise that they already have an existing account *after using the app anonymously and having already stored data in the anonymous account*.
```

> When they log in, the anonymous account will be deleted and this could lead to unexpected data loss.

Author: ```suggestion
When they now *log in*, by default the anonymous account will be discarded and this could lead to unexpected data loss.
```

> To avoid this situation we provide the `onAnonymousAccountDiscarded` handler to migrate the data from the anonymous account to the authenticated one.

Author: ```suggestion
To avoid this situation we provide the `onAnonymousAccountDiscarded` handler to migrate the data from the anonymous account to the existing authenticated one.
```

This is an example from our [music player](https://github.com/garden-co/jazz/tree/main/examples/music-player):
<CodeGroup>
```ts
export async function onAnonymousAccountDiscarded(
  anonymousAccount: MusicaAccount,
) {
  const { root: anonymousAccountRoot } = await anonymousAccount.ensureLoaded({
    root: {
      rootPlaylist: {
        tracks: [{}],
      },
    },
  });

  const me = await MusicaAccount.getMe().ensureLoaded({
    root: {
      rootPlaylist: {
        tracks: [],
      },
    },
  });

  for (const track of anonymousAccountRoot.rootPlaylist.tracks) {
    if (track.isExampleTrack) continue;

    const trackGroup = track._owner.castAs(Group);
    trackGroup.addMember(me, "admin");

    me.root.rootPlaylist.tracks.push(track);
  }
}
```
</CodeGroup>

To see how this works in reality we suggest you to try 
to upload a song in the [music player demo](https://music-demo.jazz.tools/) and then
try to log in with an existing account.

## Disable network sync for anonymous users

> You can disable network sync so you can decide when it makes sense for your app to work completely locally.

Author: ```suggestion
You can disable network sync to make your app local-only under specific circumstances.
```

> For example, you may want to give the opportunity to non-authenticated users to try your app locally first, then enable the network sync only when the user is authenticated:

Author: ```suggestion
For example, you may want to give the opportunity to non-authenticated users to try your app locally-only (incurring no sync traffic), then enable the network sync only when the user is authenticated:
```

<CodeGroup>
```tsx
<JazzProvider
  sync={{
    peer: `wss://cloud.jazz.tools/?key=${apiKey}`,
     // This makes the app work in local mode when the user is anonymous
    when: "signedUp",
  }}
>
  <App />
</JazzProvider>
```
</CodeGroup>

> You can manually control the sync by switching when between `"always"` and `"never"`.

Author: ```suggestion
For more complex behaviours, you can manually control sync by statefully switching when between `"always"` and `"never"`.
```
